android/canonical/app/src/main/java/com/google/samples/quickstart/canonical/LoginFragment.kt

```diff
@@ -0,0 +1,130 @@
+package com.google.samples.quickstart.canonical
+
+import android.content.Intent
+import android.os.Bundle
+import android.util.Log
+import androidx.fragment.app.Fragment
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import androidx.lifecycle.ViewModelProviders
+import com.google.android.gms.auth.api.signin.GoogleSignIn
+import com.google.android.gms.common.SignInButton
+import com.google.android.gms.common.api.ApiException
+import com.google.firebase.auth.GoogleAuthProvider
+
+// TODO: Rename parameter arguments, choose names that match
+// the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
+private const val ARG_PARAM1 = "param1"
+private const val ARG_PARAM2 = "param2"
+
+/**
+ * A simple [Fragment] subclass.
+ * Use the [LoginFragment.newInstance] factory method to
+ * create an instance of this fragment.
+ */
+class LoginFragment : Fragment() {
+    // TODO: Rename and change types of parameters
+    private var param1: String? = null
+    private var param2: String? = null
+    private lateinit var signInVM : SignInViewModel
+
+    private fun signIn() {
+        val signInIntent = signInVM.getGoogleSignInClient().signInIntent
+        startActivityForResult(signInIntent, RC_SIGN_IN)
```

Comment:
Instead of reaching through to the ViewModel to get GoogleSignInClient and call methods on it you could do this:

```
val signInIntent = signInVM.getSignInIntent()
startActivityForResult(signInIntent, RC_SIGN_IN)
```

The goal is that the Fragment doesn't "know" anything about signing in. It just draws pixels and does navigation by reacting to LiveData changes through an observer and calling methods on the ViewModel. The ViewModel contains all of the sign-in knowledge. With that separation of concerns, you reduce coupling and increase cohesion.

If someone comes along later to replace this fragment or add authentication to another fragment, they don't need to know how auth works or repeat any non-trivial code.